gravitee-apim-portal-webui-next/projects/gravitee-dashboard/src/lib/components/widget/widget.ts

```diff
@@ -17,12 +17,89 @@ import { GridsterItem } from 'angular-gridster2';
 
 import { PieType } from '../chart/pie-chart/pie-chart.component';
 
-export type WidgetType = PieType | 'kpi' | 'top';
+export type WidgetType = PieType | 'stats' | 'top';
 
-export interface Widget {
+export type MeasuresType = 'AVG' | 'MIN' | 'MAX' | 'P50' | 'P90' | 'P95' | 'P99' | 'COUNT' | 'RPS';
+
+export interface Widget2 {
   id: string;
   label: string;
   type: WidgetType;
   filter?: string;
   layout: GridsterItem;
 }
+
+export interface Widget {
+  id: string;
+  title: string;
+  type: WidgetType;
+  layout: WidgetLayout;
+
+  request?: MetricsRequest;
+  data?: MetricsResponse;
+}
+
+export interface MetricsResponse {
+  interval?: string;
+  metrics?: Metric[]; //remove ? after remove mock
+}
+
+export interface Metric {
+  name: string;
+  buckets?: Bucket[]; //Used by Facets and time-series
+  measures?: Measure[]; // Used by KPI/STATS
+  status?: 'pending' | 'success' | 'failed';
+}
+
+export interface Bucket {
+  key: string;
+  timestamp?: Date; //used by time-series
+  measures?: Measure[];
+  buckets?: Bucket[];
+}
+export interface Measure {
+  name: string;
+  value: number;
+}
+
+export interface WidgetLayout {
+  cols: number;
+  rows: number;
+  x: number;
+  y: number;
+}
+
+//Metrics
+
+export interface MetricsRequest {
+  type: 'measures' | 'facets' | 'time-series';
```

Comment:
What do you think about structuring the strong type here instead of adding comments and "undefine-able" properties? It would be possible to do it like in OpenAPI. What do you think?